test/cpp_extensions/open_registration_extension/torch_openreg/csrc/runtime/OpenRegDeviceAllocator.cpp

```diff
@@ -1,8 +1,205 @@
 #include "OpenRegDeviceAllocator.h"
 
+#include <c10/util/Exception.h>
+#include <c10/util/irange.h>
+
 namespace c10::openreg {
 
-static OpenRegDeviceAllocator global_openreg_alloc;
-REGISTER_ALLOCATOR(c10::DeviceType::PrivateUse1, &global_openreg_alloc);
+// Anonymous namespace for internal implementation
+namespace {
+
+// Static global allocator instance
+static OpenRegDeviceAllocator allocator;
+
+// Static deleter function
+static void deleteOpenRegMemory(void* ptr) {
+  allocator.freeMemory(ptr);
+}
+
+} // anonymous namespace
+
+OpenRegDeviceAllocator::OpenRegDeviceAllocator() {
+  // Initialize stats to zero (default constructor already does this)
+}
+
+at::DataPtr OpenRegDeviceAllocator::allocate(size_t nbytes) {
+  // Get current device first (before locking)
+  int current_device_index = -1;
+  auto ret = orGetDevice(&current_device_index);
+  TORCH_CHECK(
+      ret == orSuccess,
+      "Failed to get current OpenReg device");
+
+  auto curr_device =
+      c10::Device(c10::DeviceType::PrivateUse1, current_device_index);
+
+  void* data = nullptr;
+  if (nbytes > 0) {
+    // Lock before allocation to ensure thread-safe statistics update
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    // Allocate memory
+    ret = orMalloc(&data, nbytes);
+
+    TORCH_CHECK(
+        ret == orSuccess && data != nullptr,
+        "Failed to allocate ", nbytes, " bytes on openreg device. ",
+        "Allocated: ", stats_.allocated_bytes[0].current, " bytes, ",
+        "Reserved: ", stats_.reserved_bytes[0].current, " bytes");
+
+    // Track allocation size for proper deallocation statistics
+    allocation_sizes_[data] = nbytes;
+
+    // Update statistics
+    // StatType::AGGREGATE (index 0) is for aggregate statistics
+    using namespace c10::CachingAllocator;
+    constexpr size_t kAggregate = static_cast<size_t>(StatType::AGGREGATE);
+
+    stats_.allocated_bytes[kAggregate].increase(nbytes);
+    stats_.reserved_bytes[kAggregate].increase(nbytes);
+    stats_.num_device_alloc++;
+  }
+
+  return {data, data, &deleteOpenRegMemory, curr_device};
+}
+
+at::DeleterFnPtr OpenRegDeviceAllocator::raw_deleter() const {
+  return &deleteOpenRegMemory;
+}
+
+void OpenRegDeviceAllocator::copy_data(
+    void* dest,
+    const void* src,
+    std::size_t count) const {
+  auto ret = orMemcpy(dest, src, count, orMemcpyDeviceToDevice);
+  TORCH_CHECK(
+      ret == orSuccess,
+      "Failed to copy ", count, " bytes on openreg device");
+}
+
+// ============ DeviceAllocator Interface Implementation ============
+
+bool OpenRegDeviceAllocator::initialized() {
+  // OpenReg is always initialized (simplified implementation)
+  // TODO: When caching is implemented, release all free blocks here
+  return true;
+}
+
+void OpenRegDeviceAllocator::emptyCache(MempoolId_t mempool_id) {
+  // Phase 1: Empty implementation (no caching mechanism yet)
+  // TODO: When caching is implemented, release all free blocks here
+}
```

Comment:
please delete this. 
We can add it back later when we add caching support for OpenReg.